deploy/rbd/kubernetes/csi-rbdplugin.yaml

```diff
@@ -19,6 +19,7 @@ spec:
       # to use e.g. Rook orchestrated cluster, and mons' FQDN is
       # resolved through k8s service, set dns policy to cluster first
       dnsPolicy: ClusterFirstWithHostNet
+      terminationGracePeriodSeconds: 600
```

Comment:
> Although 600 seconds is probably too long, increasing this timeout from the default 30 seconds would certainly help in many cases. But, as mentioned in #2298, it is just a workaround and can still lead to lost I/O and user workloads getting stuck.

Thanks @idryomov 

Yes, it is also good to understand that the probability of holding IO for longer than 600 sec is very less, and there is no way to avoid this completely as far as I understand, unless we agree to wait forever, which is very impractical here IMO.

> At the risk of sounding like a broken record: with `rbd-nbd` mounter, the nodeplugin pod is no longer a "simple" CSI pod where all of the I/O logic/state is contained in the kernel. It is in the I/O path and must be treated specially. For upgrade in particular, it feels like a better approach would be to ask k8s to not schedule any new nodeplugin-dependent pods on the node, gracefully drain existing or already scheduled nodeplugin-dependent pods and then upgrade the nodeplugin pod. This seems less error prone and more general -- for example, it would also work for `ceph-fuse` which doesn't have an equivalent of `rbd-nbd attach` (at least AFAIK).

Yes. The rbd nodeplugin is a daemon-set and hence with the daemon-sets upgrade strategy the ordering of old and new pods going down and coming up respectively will be promising.

```
DaemonSet Update Strategy

DaemonSet has two update strategy types:

OnDelete:
With OnDelete update strategy, after you update a DaemonSet template, new DaemonSet pods will
only be created when you manually delete old DaemonSet pods. This is the same behavior of
DaemonSet in Kubernetes version 1.5 or before.

RollingUpdate:
This is the default update strategy.
With RollingUpdate update strategy, after you update a DaemonSet template, old DaemonSet pods
will be killed, and new DaemonSet pods will be created automatically, in a controlled fashion. At most
one pod of the DaemonSet will be running on each node during the whole update process.
```

Reply:
Yes, `OnDelete` would be part of what I'm thinking about.  It looks like most of this is already documented at https://github.com/ceph/ceph-csi/blob/devel/docs/ceph-csi-upgrade.md but currently it is a manual process.  Has any thought been put into automating this in some higher-level operator (e.g. OCS operator)?

I mean, upgrade the most important (perhaps even the only?) use case we are chasing here, right?  We can't do much about an ungraceful restart of the nodeplugin pod, but automating upgrades seems feasible to me.

Reply:
I'm not up to date about the automation effort.
@Madhu-1 @nixpanic can you please chime in?

Reply:
we have seen issues where PodNetworking (as opposed to HostNetworking) is a problem. As that changes the (Linux) network namespace of the connections, restarting the ceph-csi pods might prevent further operations on the mountpoints (basically prevents unmounting, I thought). For this to work out, updates might require to drain all Pods that use Ceph-CSI volumes from the node where the update is taking place. To me it is unclear what component should be coordinating the update of Ceph-CSI, often it is Rook that deploys Ceph-CSI, so that is the main candidate from my understanding.

Draining application Pods before updating is quite a big hammer, but it would be an approach that works for this as well.

Reply:
Yes, our recent troubles with pod networking is why I brought this up.  While tweaking timeouts might help `rbd-nbd` in this specific case, this is a general problem and we need a general solution.  I agree that draining is a quite big hammer, but it seems to be the only approach that is in line with k8s.

That said I'm not really a k8s person, so I may be missing something.  Just trying to avoid the pile up of ad-hoc workarounds that don't really solve the problem...

Reply:
@idryomov

In a short discussion with @nixpanic we came across the below idea, and we need your thoughts on this :-)

On the other comment, Niels thinks that the data is probably not reaching the rbd-nbd maybe it might be possible that writes are cached at filesystem or in the block layer too, that way issuing the sync from the mount might help push them to rbd-nbd tray at least. But again `sync ` might be possible to exclusively run in our e2e, but we cannot ensure that to happen with real workloads. For this, to happen probably we should pass `-o sync` at mount time, but that will reduce the IO performance drastically.

We are curious to know, just like the quiesce hook which executes `fsfreeze` on the mount before starting to take a snapshot of an image and then `fsfreeze -u` after the snapshot is taken, can we run the quiesce hook for attach and detach operations? i.e as part of the quiescing, call `fsfreeze` before the detach is called, and then allow reattach the process then invoke `fsfreeze -u`?

We understand that the freezing + flushing of the pending inflight IO takes time too, but we should be in a better place to ensure that the data is not cached/waiting at the other layers. So if that causes pods to get terminated before flushing data is complete at rbd-nbd, there is a good reason to extend the default pod termination grace timeout (or use the application lifecycle hooks)